testing/integration-tests/src/codegen/codegen_tests.rs

```diff
@@ -0,0 +1,97 @@
+    let item_mod = syn::parse_quote!(
+        pub mod api {}
+    );
+    let crate_path = CratePath::default();
+    let derives = DerivesRegistry::with_default_derives(&crate_path);
+    let type_substitutes = TypeSubstitutes::with_default_substitutes(&crate_path);
+    generator
+        .generate_runtime(item_mod, derives, type_substitutes, crate_path, false)
+        .expect("API generation must be valid")
+        .to_string()
+}
+
+#[test]
+fn dupe_types_do_not_overwrite_each_other() {
+    fn generate_metadata_with_duplicate_types() -> RuntimeMetadataPrefixed {
+        #[derive(TypeInfo)]
+        struct Runtime;
+        #[derive(TypeInfo)]
+        enum RuntimeCall {}
+        #[derive(TypeInfo)]
+        enum RuntimeEvent {}
+        #[derive(TypeInfo)]
+        pub enum DispatchError {}
+
+        // We need these types for codegen to work:
+        let mut registry = scale_info::Registry::new();
+        let ty = registry.register_type(&meta_type::<Runtime>());
+        registry.register_type(&meta_type::<RuntimeCall>());
+        registry.register_type(&meta_type::<RuntimeEvent>());
+        registry.register_type(&meta_type::<DispatchError>());
+
+        // Now we duplicate some types with same type info:
+        enum Foo {}
+        impl TypeInfo for Foo {
+            type Identity = Self;
+            fn type_info() -> scale_info::Type {
+                scale_info::Type::builder()
+                    .path(scale_info::Path::new("DuplicateType", "dupe_mod"))
+                    .variant(
+                        scale_info::build::Variants::new()
+                            .variant("FirstDupeTypeVariant", |builder| builder.index(0)),
+                    )
+            }
+        }
+        enum Bar {}
+        impl TypeInfo for Bar {
+            type Identity = Self;
+            fn type_info() -> scale_info::Type {
+                scale_info::Type::builder()
+                    .path(scale_info::Path::new("DuplicateType", "dupe_mod"))
+                    .variant(
+                        scale_info::build::Variants::new()
+                            .variant("SecondDupeTypeVariant", |builder| builder.index(0)),
+                    )
+            }
+        }
+
+        registry.register_type(&meta_type::<Foo>());
+        registry.register_type(&meta_type::<Bar>());
+
+        let extrinsic = ExtrinsicMetadata {
+            ty: meta_type::<()>(),
+            version: 0,
+            signed_extensions: vec![],
+        }
+        .into_portable(&mut registry);
+        let metadata = RuntimeMetadataV15 {
+            types: registry.into(),
+            pallets: Vec::new(),
+            extrinsic,
+            ty,
+            apis: vec![],
+        };
+
+        RuntimeMetadataPrefixed::from(metadata)
+    }
+
+    let metadata = generate_metadata_with_duplicate_types();
+    let interface = generate_runtime_interface_from_metadata(metadata);
+
+    assert!(interface.contains("FirstDupeTypeVariant"));
```

Comment:
see my comment above, I don't understand if the the variants are merged to a single type or something else?

Reply:
We end up with 2 variant types in the registry that both have the same path and name, and then when doing the codegen we spot that the paths are identical and add a "2" to one of the names to make sure that they don't override eachother. (ie if you remove the `ensure_unique_types` line and run this test again it'll fail)